public/app/features/dashboard/components/DashboardPrompt/DashboardPrompt.tsx

```diff
@@ -0,0 +1,217 @@
+import { locationService } from '@grafana/runtime';
+import { appEvents } from 'app/core/core';
+import { contextSrv } from 'app/core/services/context_srv';
+import React, { useEffect, useState } from 'react';
+import { Prompt } from 'react-router-dom';
+import { DashboardModel } from '../../state/DashboardModel';
+import { each, filter, find } from 'lodash';
+import angular from 'angular';
+import { UnsavedChangesModal } from '../SaveDashboard/UnsavedChangesModal';
+import * as H from 'history';
+import { SaveLibraryPanelModal } from 'app/features/library-panels/components/SaveLibraryPanelModal/SaveLibraryPanelModal';
+import { PanelModelWithLibraryPanel } from 'app/features/library-panels/types';
+import { useDispatch } from 'react-redux';
+import { discardPanelChanges } from '../PanelEditor/state/actions';
+import { DashboardSavedEvent } from 'app/types/events';
+
+export interface Props {
+  dashboard: DashboardModel;
+}
+
+interface State {
+  original: object | null;
+  originalPath?: string;
+  modal: PromptModal | null;
+  blockedLocation?: H.Location | null;
+}
+
+enum PromptModal {
+  UnsavedChangesModal,
+  SaveLibraryPanelModal,
+}
+
+export const DashboardPrompt = React.memo(({ dashboard }: Props) => {
+  const [state, setState] = useState<State>({ original: null, modal: null });
+  const dispatch = useDispatch();
+  const { original, originalPath, blockedLocation, modal } = state;
+
+  useEffect(() => {
+    const timeoutId = setTimeout(() => {
+      const originalPath = locationService.getLocation().pathname;
+      const original = dashboard.getSaveModelClone();
+
+      setState({ originalPath, original, modal: null });
+    }, 1000);
+
+    const savedEventUnsub = appEvents.subscribe(DashboardSavedEvent, () => {
+      const original = dashboard.getSaveModelClone();
+      const originalPath = locationService.getLocation().pathname;
+      setState({ originalPath, original, modal: null });
+    });
+
+    return () => {
+      clearTimeout(timeoutId);
+      savedEventUnsub.unsubscribe();
+    };
+  }, [dashboard]);
+
+  const onHistoryBlock = (location: H.Location) => {
+    const panelInEdit = dashboard.panelInEdit;
+    const search = new URLSearchParams(location.search);
+
+    // Are we leaving panel edit & library panel?
+    if (panelInEdit && panelInEdit.libraryPanel && panelInEdit.hasChanged && !search.has('editPanel')) {
```

Comment:
Why do we have a separate check for the library panel?

Reply:
@dprokop because we have a separate unsaved change modal for unsaved library panel changes when leaving panel edit (the main reason for this PR was that we had a history.block Prompt in panel edit and it overwrote the one in ChangeTracker)

Reply:
ok, I see, thanks!